### ISSUE TITLE
[BUG] named bind/mount failing since newer ubuntu/docker version

### ISSUE THREAD

**Issue by typoworx-de** [docker/compose#11678]
### Description

I've recently upgraded my Ubuntu to recent 23.10 and I think docker version is newer as well. A bunch of docker-compose.yaml is failing with errors like this:
```Error response from daemon: failed to mount local volume: mount /var/www/self-hosted/:/var/snap/docker/common/var-lib-docker/volumes/self-hosted_web-root/_data, flags: 0x1000, data: uid=33,gid=33: no such file or directory```

This is the part in the docker-compose.yaml which seems to cause the problem:
```
volumes:
  web-root:
    driver: local
    driver_opts:
      type: none
      device: "/var/www/self-hosted/"
      o: "bind,uid=${APP_UID:-33},gid=${APP_UID:-33}"
```

And yes I could use direct mount, but I would like to make shure the mount is owned by www-data (uid/gid 33) which at least worked fine before upgrading my linux-system and docker.

Of course I could always modify/extend a bunch of docker-images making a new entrypoint which will fix this, but I think there should be a better solution for this instead of such poking with workarounds.

So why isn't this working anymore?! 

Issue sounds similar to:
#7791

### Steps To Reproduce

1. create docker-compose.yaml with named volume bind/mount to path
2. create any service using that named volume
3. try to start the stack using 'docker-compose up -d'

### Compose Version

```Text
Docker Compose version v2.25.0
```

### Docker Environment

```Text
Client:
 Version:    24.0.5
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.2
    Path:     /usr/local/lib/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.25.0
    Path:     /usr/local/lib/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 2
  Running: 1
  Paused: 0
  Stopped: 1
 Images: 21
 Server Version: 24.0.5
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dce8eb055cbb6872793272b4f20ed16117344f8
 runc version: 
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.5.0-10027-tuxedo
 Operating System: Ubuntu Core 22
 OSType: linux
 Architecture: x86_64
 CPUs: 32
 Total Memory: 62.53GiB
 Name: Gabriel-Tuxedo
 ID: 7c4b6771-e77e-445c-b707-e70f07913eac
 Docker Root Dir: /var/snap/docker/common/var-lib-docker
 Debug Mode: false
 Username: typoworx
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by typoworx-de** [comment]
Upgrading to more recent docker-compose v2.26.1 didn't solve the issue.

**Comment by typoworx-de** [comment]
I think I've found the culprit in the meantime. For some reason Ubuntu installed docker using snap(!!!) :warning: 
which caused massive problems.

I installed docker myself using apt from latest regular repository so there where two docker's installed and for some reason sometimes the damn snapped docker has taken over control or where running containers even while my regular dockerd was stopped.

I not even noticed nor installed the snap docker, so I think this was the problem at all. I'm closing this issue for know as my containers start now without issues. I think the snapped docker hat to restricted privileges which caused the problems at all!
